aclk/aclk_query.c

```diff
@@ -0,0 +1,574 @@
+            meta_state = aclk_shared_state.metadata_submitted;
+            aclk_shared_state.metadata_submitted = ACLK_METADATA_SENT;
+            ACLK_SHARED_STATE_UNLOCK;
+            aclk_send_metadata(meta_state);
+            break;
+
+        case ACLK_CMD_CHART:
+            debug(D_ACLK, "EXECUTING a chart update command");
+            aclk_send_single_chart(this_query->data, this_query->query);
+            break;
+
+        case ACLK_CMD_CHARTDEL:
+            debug(D_ACLK, "EXECUTING a chart delete command");
+            //TODO: This send the info metadata for now
+            aclk_send_info_metadata(ACLK_METADATA_SENT);
+            break;
+
+        case ACLK_CMD_ALARM:
+            debug(D_ACLK, "EXECUTING an alarm update command");
+            aclk_send_message(this_query->topic, this_query->query, this_query->msg_id);
+            break;
+
+        case ACLK_CMD_CLOUD:
+            t = now_monotonic_high_precision_usec();
+            debug(D_ACLK, "EXECUTING a cloud command");
+            aclk_execute_query(this_query);
+            t = now_monotonic_high_precision_usec() - t;
+            break;
+
+        default:
+            break;
+    }
+    debug(D_ACLK, "Query #%ld (%s) done", query_count, this_query->topic);
+
+    if (aclk_stats_enabled) {
+        ACLK_STATS_LOCK;
+        aclk_metrics_per_sample.queries_dispatched++;
+        aclk_queries_per_thread[t_idx]++;
+        if(this_query->cmd == ACLK_CMD_CLOUD) {
+            aclk_metrics_per_sample.cloud_q_process_total += t;
+            aclk_metrics_per_sample.cloud_q_process_count++;
+            if(aclk_metrics_per_sample.cloud_q_process_max < t)
+                aclk_metrics_per_sample.cloud_q_process_max = t;
+        }
+        ACLK_STATS_UNLOCK;
+    }
+
+    aclk_query_free(this_query);
+
+    return 1;
+}
+
+void aclk_query_threads_cleanup(struct aclk_query_threads *query_threads)
+{
+    if (query_threads) {
+        for (int i = 0; i < query_threads->count; i++) {
+            netdata_thread_join(*(query_threads->thread_list[i]->thread), NULL);
+            freez(query_threads->thread_list[i]->thread);
+            freez(query_threads->thread_list[i]);
+        }
+    }
+
+    struct aclk_query *this_query;
+
+    do {
+        this_query = aclk_queue_pop();
+        aclk_query_free(this_query);
+    } while (this_query);
+}
+
+#define TASK_LEN_MAX 16
+void aclk_query_threads_start(struct aclk_query_threads *query_threads)
+{
+    info("Starting %d query threads.", query_threads->count);
+
+    char thread_name[TASK_LEN_MAX];
+    query_threads->thread_list = callocz(query_threads->count, sizeof(struct aclk_query_thread *));
+    for (int i = 0; i < query_threads->count; i++) {
+        query_threads->thread_list[i] = callocz(1, sizeof(struct aclk_query_thread));
+        query_threads->thread_list[i]->thread = callocz(1, sizeof(netdata_thread_t));
+        query_threads->thread_list[i]->idx = i; //thread needs to know its index for statistics
```

Comment:
Do we need so many pointers? I'm having difficulty following all the memory allocations related to threads. Could this work if it was like the following?
```
    query_threads->thread_list = callocz(query_threads->count, sizeof(struct aclk_query_thread));
    for (int i = 0; i < query_threads->count; i++) {
        query_threads->thread_list[i].thread = callocz(1, sizeof(netdata_thread_t));
        query_threads->thread_list[i].idx = i; //thread needs to know its index for statistics 

        snprintf(thread_name, TASK_LEN_MAX, "%s_%d", ACLK_THREAD_NAME, i);
        netdata_thread_create(
            query_threads->thread_list[i].thread, thread_name, NETDATA_THREAD_OPTION_JOINABLE, aclk_query_main_thread,
            query_threads->thread_list[i]);
    } 
```
Moreover, could the `netdata_thread_t` be part of `struct aclk_query_thread` thus removing the extra allocation?
```
    query_threads->thread_list = callocz(query_threads->count, sizeof(struct aclk_query_thread));
    for (int i = 0; i < query_threads->count; i++) {
        query_threads->thread_list[i].idx = i; //thread needs to know its index for statistics 

        snprintf(thread_name, TASK_LEN_MAX, "%s_%d", ACLK_THREAD_NAME, i);
        netdata_thread_create(
            &query_threads->thread_list[i].thread, thread_name, NETDATA_THREAD_OPTION_JOINABLE, aclk_query_main_thread,
            query_threads->thread_list[i]);
    } 
```

Reply:
Thanks, you are right that is unnecessary indirection and only makes code harder to read. I simplified it.